docs(expo-router): document server rendering (SSR) (#42266)
expo/expo: docs/pages/router/web/server-rendering.mdx

---
title: Server Rendering
description: Learn how to render Expo Router routes dynamically at request time using server-side rendering (SSR).

> isNew: true

Author: ```suggestion
```

Author: As per @amandeepmittal's suggestion on my other PR, `isNew` should not be used for Guides

{
  "expo": {
    /* @hide ... */
    /* @end */
    "web": {
      /* @info Server rendering requires Metro bundler */
      "bundler": "metro",
      /* @end */
      "output": "server"
    }
  }
}
```

</Step>

<Step label="2">

Enable server rendering in the `expo-router` plugin configuration:

```json app.json
{
  "expo": {
    /* @hide ... */
    /* @end */
    "plugins": [
      [
        "expo-router",
        {
          /* @info Enable server-side rendering */
          "unstable_useServerRendering": true
          /* @end */
        }
      ]
    ]
  }
}
```

</Step>

<Step label="3">

If you have a **metro.config.js** file in your project, ensure it extends `expo/metro-config`:

```js metro.config.js
const { getDefaultConfig } = require('expo/metro-config');

/** @type {import('expo/metro-config').MetroConfig} */
const config = getDefaultConfig(__dirname, {
  // Additional features...
});

module.exports = config;
```

</Step>

<Step label="4">

Start the development server:

<Terminal cmd={['$ npx expo start']} />

</Step>

## Production

To export your app for production, run the universal export command:

<Terminal cmd={['$ npx expo export --platform web']} />

> This creates a **dist** directory with your server-rendered application. Unlike static rendering, no HTML files are pre-generated. Instead, the output includes:

Reviewer: ```suggestion
This creates a **dist** directory with your server-rendered application. Unlike static rendering, no HTML files are pre-generated. Instead, the output includes a similar directory structure as shown below:
```

Let's clarify this.

        {/*
          Disable body scrolling on web. This makes ScrollView components work closer to how they do on native.
          However, body scrolling is often nice to have for mobile web. If you want to enable it, remove this line.
        */}
        <ScrollViewStyleReset />

        {/* Add any additional <head> elements that you want globally available on web... */}
      </head>
      <body>{children}</body>
    </html>
  );
}
```

The **+html.tsx** file:

- Runs only in Node.js during server rendering
- Cannot import global CSS (use the [Root Layout](/router/basics/layout/#root-layout) for styles)
- Cannot access browser APIs like `window` or `document`
- Receives pre-rendered content via the `children` prop

## Meta tags

Add meta tags to your pages using the `<Head />` component from `expo-router`:

```tsx app/about.tsx
import Head from 'expo-router/head';
import { Text } from 'react-native';

export default function Page() {
  return (
    <>
      <Head>
        <title>About Us</title>
        <meta name="description" content="Learn more about our company." />
      </Head>
      <Text>About page content</Text>
    </>
  );
}
```

> In server rendering mode, `<Head>` elements are rendered server-side and included in the initial HTML response, improving search engine optimization (SEO).

Reviewer 2: Let's avoid saying "server rendering mode" and instead prefer "server-side rendering", since we rarely have to call out the mode, and this may be confused with the previous "server output mode"

```suggestion
During server-side rendering, `<Head>` elements are extracted and included in the initial HTML response, modifying the `<head>` element sent to the client, and improving search engine optimization (SEO).
```

## Deployment

Server rendering requires a runtime server to render pages on each request. You **cannot** deploy to static hosting services like GitHub Pages.

### Supported platforms

| Platform                                  | Adapter                       |
| ----------------------------------------- | ----------------------------- |
| [EAS Hosting](/eas/hosting/introduction/) | Built-in                      |
| Node.js/Express                           | `expo-server/adapter/express` |
| Cloudflare Workers                        | `expo-server/adapter/workerd` |
| Vercel Edge Functions                     | `expo-server/adapter/vercel`  |
| Netlify Edge Functions                    | `expo-server/adapter/netlify` |
| Bun                                       | `expo-server/adapter/bun`     |

> <Collapsible summary="Example: Deploying to EAS Hosting">

Reviewer: ```suggestion
<Collapsible summary="Example: Deployment with EAS Hosting">
```
